src/aiai_eval/named_entity_recognition.py

```diff
@@ -251,35 +365,67 @@ def _prepare_predictions_and_labels(
-            predictions = [
-                [
-                    id2label[pred_id]
-                    for pred_id, lbl_id in zip(pred, label)
-                    if lbl_id != -100
+        # In case we have received predictions and labels which are already converted
+        # from ids to label tags, we do not need to convert them again.
+        # This is the case when the model is a spacy model.
+        if any(
+            predictions_np.dtype == dtype
+            for dtype in {
+                np.float16,
+                np.float32,
+                np.float64,
+                np.int16,
+                np.int32,
+                np.int64,
+            }
+        ) and any(
+            labels_np.dtype == dtype
+            for dtype in {
+                np.float16,
+                np.float32,
+                np.float64,
+                np.int16,
+                np.int32,
+                np.int64,
+            }
+        ):
+
+            if id2label is not None:
+                # Remove ignored index (special tokens)
+                predictions = [
+                    [
+                        id2label[pred_id]
+                        for pred_id, lbl_id in zip(pred, label)
+                        if lbl_id != -100
+                    ]
+                    for pred, label in zip(predictions_np, labels_np)
+                ]
+                labels = [
+                    [
+                        id2label[lbl_id]
+                        for _, lbl_id in zip(pred, label)
+                        if lbl_id != -100
+                    ]
+                    for pred, label in zip(predictions_np, labels_np)
                 ]
-                for pred, label in zip(predictions_np, labels_np)
-            ]
-            labels = [
-                [id2label[lbl_id] for _, lbl_id in zip(pred, label) if lbl_id != -100]
-                for pred, label in zip(predictions_np, labels_np)
-            ]
 
-        # Replace predicted tag with either MISC or O tags if they are not part of the
-        # dataset
-        id2label_without_misc = set(self.task_config.id2label).difference(
-            {"B-MISC", "I-MISC"}
-        )
-        for i, prediction_list in enumerate(predictions):
-            for j, ner_tag in enumerate(prediction_list):
-                if ner_tag not in id2label_without_misc:
-                    if self.has_misc_tags and id2label[ner_tag][:2] == "B-":  # type: ignore
-                        predictions[i][j] = "B-MISC"
-                    elif self.has_misc_tags and id2label[ner_tag][:2] == "I-":  # type: ignore
-                        predictions[i][j] = "I-MISC"
-                    else:
-                        predictions[i][j] = "O"
+            # Replace predicted tag with either MISC or O tags if they are not part of the
+            # dataset
+            id2label_without_misc = set(self.task_config.id2label).difference(
+                {"B-MISC", "I-MISC"}
+            )
```

Comment:
Using the dataset `id2label` will cause a mismatch for transformer-based models with different `id2label` mappings, no?